### PR TITLE
tailcfg: add NodeCapMap

### PR DESCRIPTION
**Pull request by maisem**
Like PeerCapMap, add a field to `tailcfg.Node` which provides a map of Capability to raw JSON messages which are deferred to be parsed later by the application code which cares about the specific capabilities. This effectively allows us to prototype new behavior without having to commit to a schema in tailcfg, and it also opens up the possibilities to develop custom behavior in tsnet applications w/o having to plumb through application specific data in the MapResponse.

Updates #4217

### CODE REVIEW

File: tailcfg/tailcfg.go [hunk 1]
```diff
@@ -292,6 +317,9 @@ type Node struct {
 	//    "https://tailscale.com/cap/file-sharing"
 	Capabilities []NodeCapability `json:",omitempty"`
 
+	// CapMap is a map of capabilities to their raw values.
+	CapMap NodeCapMap
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:321]: omitempty?

File: tailcfg/tailcfg.go [hunk 2]
```diff
@@ -343,6 +371,16 @@ type Node struct {
 	ExitNodeDNSResolvers []*dnstype.Resolver `json:",omitempty"`
 }
 
+// HasCap reports whether the node has the given capability.
+func (v NodeView) HasCap(cap NodeCapability) bool {
+	return v.ж.HasCap(cap)
+}
+
+// HasCap reports whether the node has the given capability.
+func (v *Node) HasCap(cap NodeCapability) bool {
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:380]: HasCap here vs HasCapability on the other type is weird. pick a name?

Comment (bradfitz) [reply tailcfg/tailcfg.go:380]: Alternative, HasCap here but just "Contains" on the other one, since it's only about capabilities it doesn't need "Cap" or "Capabilities" in its name and "Contains" seems like the common name.

File: tailcfg/tailcfg.go [hunk 3]
```diff
@@ -1250,13 +1288,64 @@ const (
...
+	if len(c) != len(c2) {
+		return false
+	}
+	for k, v := range c {
+		v2, ok := c2[k]
+		if !ok {
+			return false
+		}
+		if !slices.EqualFunc(v, v2, func(e1, e2 RawMessage) bool {
+			return bytes.Equal([]byte(e1), []byte(e2))
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:1307]: why not e1 == e2?

Comment (maisem) [reply tailcfg/tailcfg.go:1307]: Oh yeah that'll work too

File: tailcfg/tailcfg.go [hunk 4]
```diff
@@ -1250,13 +1288,64 @@ const (
...
+			return nil, err
+		}
+		out = append(out, t)
+	}
+	return out, nil
+}
+
+// HasCapability reports whether c has the capability cap.
+// This is used to test for the existence of a capability, especially
+// when the capability has no values.
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:1336]: maybe reword "has no values" somehow to be more clear?

"has no associated argument/data values"
"has no ancillary values"

We should come up with a good name for that optional value bit and use its name consistently.

File: ipn/ipnlocal/peerapi.go [hunk 5]
```diff
@@ -1035,7 +1035,7 @@ func (h *peerAPIHandler) canPutFile() bool {
 // canDebug reports whether h can debug this node (goroutines, metrics,
 // magicsock internal state, etc).
 func (h *peerAPIHandler) canDebug() bool {
-	if !views.SliceContains(h.selfNode.Capabilities(), tailcfg.CapabilityDebug) {
+	if !h.selfNode.HasCap(tailcfg.CapabilityDebug) {
```
Comment (bradfitz) [inline ipn/ipnlocal/peerapi.go:1038]: all these changes could be in their own commit if you wanted to make this smaller.

File: control/controlknobs/controlknobs.go [hunk 6]
```diff
@@ -48,39 +49,30 @@ type Knobs struct {
...
-		case tailcfg.NodeAttrDebugForceBackgroundSTUN:
-			forceBackgroundSTUN = true
-		case tailcfg.NodeAttrDisableDeltaUpdates:
-			disableDeltaUpdates = true
-		}
+
+	if has(tailcfg.NodeAttrOneCGNATEnable) {
+		oneCGNAT.Set(true)
+	} else if has(tailcfg.NodeAttrOneCGNATDisable) {
+		oneCGNAT.Set(false)
```
Comment (bradfitz) [inline control/controlknobs/controlknobs.go:73]: this is a bit worse performance-wise now. Now it does 8 passes over the slice.

Comment (maisem) [reply control/controlknobs/controlknobs.go:73]: I expect we will just start sending down the CapMap from control instead.

File: tailcfg/tailcfg.go [hunk 7]
```diff
@@ -368,6 +368,16 @@ type Node struct {
 	ExitNodeDNSResolvers []*dnstype.Resolver `json:",omitempty"`
 }
 
+// HasCap reports whether the node has the given capability.
+func (v NodeView) HasCap(cap NodeCapability) bool {
+	return v.ж.HasCap(cap)
+}
+
+// HasCap reports whether the node has the given capability.
+func (v *Node) HasCap(cap NodeCapability) bool {
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:381]: document that both of these are safe to call on a nil/invalid value?

File: tailcfg/tailcfg.go [hunk 8]
```diff
@@ -317,6 +318,9 @@ type Node struct {
 	//    "https://tailscale.com/cap/file-sharing"
 	Capabilities []NodeCapability `json:",omitempty"`
 
+	// CapMap is a map of capabilities to their raw values.
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:321]: say more about the values and how they can be nil? and the relation between this and the Capabilities slice?

File: tailcfg/tailcfg.go [hunk 9]
```diff
@@ -1285,6 +1289,57 @@ const (
...
+	for k, v := range c {
+		v2, ok := c2[k]
+		if !ok {
+			return false
+		}
+		if !slices.EqualFunc(v, v2, func(e1, e2 RawMessage) bool {
+			return e1 == e2
+		}) {
+			return false
+		}
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:1311]: Is this different from https://pkg.go.dev/slices#Equal ?

**Review by bradfitz — APPROVED** [review]
